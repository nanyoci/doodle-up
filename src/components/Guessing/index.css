.guessing-container {
    position: relative;
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-around;
    
}

#guessTheName{
    font-family: 'Merriweather Sans', sans-serif;
    font-size: 2vw;
    font-weight: semi-bold;
    height: 5vh;
    margin: 0px;
}

#guessingImgDiv {
    padding-top: 10px;
    position: relative;
    width: 33%;
    height: 33%;
}
  
#guessingImg{
    display: block;
  margin-left: auto;
  margin-right: auto;
    max-width:100%;
    max-height:100%;
}

.options {
    position: relative;
    width: 70%;
    display: grid;
    grid-template-columns: 1fr 1fr;
    justify-content: space-around;
    text-align: center;
}

.reward-container{
    display: flex;
    justify-content: space-between;
}

.rewards{
    overflow: hidden;
}
