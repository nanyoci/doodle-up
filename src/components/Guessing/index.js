import React, { useEffect, useState } from 'react';

import Page from '../Page';
import './index.css'


import dino from './../../assets/dino.jpg'
import audio from './../../assets/audio.svg'
import next from './../../assets/nextArrow.svg'


function Guessing() {


    const optionState = {
        initial: "btn btn-outline-dark ",
        wrong: "btn btn-danger ",
        correct: "btn btn-primary "
    }

    const [currentState, setCurrentState] = useState(
        [
            {
                state: optionState.initial
            },
            {
                state: optionState.initial
            },
            {
                state: optionState.initial
            },
            {
                state: optionState.initial
            }
        ]
    )


    const submit = (answer) => {
        let allOptions = [...currentState];
        let selectedOption = { ...allOptions[answer] };
        if (answer != 3) {
            selectedOption.state = optionState.wrong;
        }
        else {
            selectedOption.state = optionState.correct;
        }
        allOptions[answer] = selectedOption;
        setCurrentState(allOptions);
        console.log(allOptions)
    }

    return (
        <Page>
            <div id="guessing">
                <h1 id="guessTheName">Guess the Name!</h1>
                <div id="guessingImgDiv" >
                    <img id="guessingImg" src={dino} />
                </div>
                <div id="options">
                    <div className="options-row">
                        <div className="optionUnit">
                            <img src={audio} className="audioIcon" alt="audio" />
                            <button type="button" className={`${currentState[0].state} option`} disabled={currentState[0].state == optionState.wrong} onClick={() => submit(0)}>Lady Bug</button>

                        </div>
                        <div className="optionUnit">
                            <img src={audio} className="audioIcon" alt="audio" />
                            <button type="button" className={`${currentState[1].state} option`} disabled={currentState[1].state == optionState.wrong} onClick={() => submit(1)}>Scary Cat</button></div>

                    </div>
                    <div className="options-row">
                        <div className="optionUnit">
                            <img src={audio} className="audioIcon" alt="audio" />
                            <button type="button" className={`${currentState[2].state} option`} disabled={currentState[2].state == optionState.wrong} onClick={() => submit(2)}>Angry Bird</button>

                        </div>
                        <div className="optionUnit">
                            <img src={audio} className="audioIcon" alt="audio" />
                            <button type="button" className={`${currentState[3].state} option`} disabled={currentState[3].state == optionState.correct} onClick={() => submit(3)}>Dinosaur</button>

                        </div>
                    </div>
                </div>
                <img src={next} alt="next" id="next" />
            </div >
        </Page>
    );
}

export default Guessing;