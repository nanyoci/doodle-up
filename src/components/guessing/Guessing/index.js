import React, { useState, useRef } from 'react';
import useSound from 'use-sound';
import Reward from 'react-rewards';

import Page from '../../common/Page';
import NextButton from '../../common/NextButton';
import GuessingOption from '../GuessingOption';
import confettiSound from '../../../assets/soundFX/confetti.mp3';
import tryAgainSound from '../../../assets/soundFX/tryAgain.mp3';
import './index.css';

// TODO: Fix button size on ipad
function Guessing(props) {
    const {
        stage: {
            options,
            answer: answerText,
            image,
        },
        onComplete,
    } = props;

    const [isComplete, setIsComplete] = useState(false)

    const rewardLeft = useRef(null);
    const rewardRight = useRef(null);

    const [playConfettiSound] = useSound(
        confettiSound,
        { volume: 0.25 }
    )

    const [playTryAgainSound] = useSound(
        tryAgainSound,
        { volume: 0.5 }
    )

    const handleSubmit = isCorrect => {
        if (isCorrect) {
            setIsComplete(true);
            handleReward();
            playConfettiSound();
        } else {
            playTryAgainSound();
        }
    }

    const handleReward = () => {
        if (!props.test) {
            rewardLeft.current.rewardMe()
            rewardRight.current.rewardMe()
        }
    }

    return (
        <Page>
            <div className="guessing-container">
                <h1 id="guessTheName">Guess the Name!</h1>
                <div id="guessingImgDiv" >
                    <img id="guessingImg" src={image} alt="For guessing" />
                </div>
                <div className="options">
                    {
                        Object.keys(options).map(text => (
                            <GuessingOption
                                key={text}
                                text={text}
                                audio={options[text]}
                                isCorrect={text === answerText}
                                isComplete={isComplete}
                                onSubmit={handleSubmit}
                            />
                        ))
                    }
                </div>
                {
                    isComplete &&
                    <NextButton onClick={onComplete} />
                }
            </div >
            <div className='reward-container'>
                <div className='rewards'>
                    <Reward
                        ref={rewardLeft}
                        type='confetti'
                        config={
                            { angle: 50, spread: 100, elementCount: 100, elementSize: 15 }
                        }
                    >
                        {""}
                    </Reward>
                </div>
                <div className='rewards'>
                    <Reward
                        ref={rewardRight}
                        type='confetti'
                        config={
                            { angle: 130, spread: 100, elementCount: 100, elementSize: 15 }
                        }
                    >
                        {""}
                    </Reward>
                </div>
            </div>
        </Page>
    );
}

export default Guessing;