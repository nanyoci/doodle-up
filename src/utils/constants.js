export const API_URL = 'https://doodle-up-backend.herokuapp.com';