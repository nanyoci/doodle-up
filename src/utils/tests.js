import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { render } from '@testing-library/react';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )

    const renderedComponent = <Provider store={store}>{component}</Provider>;

    return {
        ...render(renderedComponent),
        store
    };
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    const renderedComponent = <Router history={history}>{component}</Router>;

    return {
        ...render(renderedComponent),
        history
    };
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    const history = createMemoryHistory({ initialEntries: [route] });

    const renderedComponent = (
        <Provider store={store}>
            <Router history={history}>{component}</Router>
        </Provider>
    );

    return {
        ...render(renderedComponent),
        history,
        store
    };
}