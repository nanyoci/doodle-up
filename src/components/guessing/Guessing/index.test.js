import React from 'react'
import { cleanup, fireEvent, waitForElement, act, screen } from '@testing-library/react'

import Guessing from './index.js';
import fakeImage from './../../../assets/drawing-sample.png';
import { renderWithReduxRouter } from './../../../utils/tests';

jest.mock('axios');


beforeEach(() => {
    window.HTMLMediaElement.prototype.load = () => { /* do nothing */ };
    window.HTMLMediaElement.prototype.play = () => { /* do nothing */ };
    window.HTMLMediaElement.prototype.pause = () => { /* do nothing */ };
})

afterEach(() => {
    cleanup();
    jest.clearAllMocks();
});

const stage = {
    options: {
        "Bear": "https://firebasestorage.googleapis.com/v0/b/doodleup-f1847.appspot.com/o/Content%2F001%2Fbear.mp3?alt=media",
        "Bed": "https://firebasestorage.googleapis.com/v0/b/doodleup-f1847.appspot.com/o/Content%2F001%2Fbed.mp3?alt=media",
        "Chair": "https://firebasestorage.googleapis.com/v0/b/doodleup-f1847.appspot.com/o/Content%2F001%2Fchair.mp3?alt=media",
        "Porridge": "https://firebasestorage.googleapis.com/v0/b/doodleup-f1847.appspot.com/o/Content%2F001%2Fporridge.mp3?alt=media"
    },
    answer: "Chair",
    image: fakeImage,
}

it('should take a snapshot', async () => {
    let asFragment = {};
    await act(async () => { ({ asFragment } = renderWithReduxRouter(<Guessing stage={stage} />)) });
    expect(asFragment()).toMatchSnapshot();
})

it('displays correct ui after correct answer is selected', async () => {
    const { getByTestId, getByText } = renderWithReduxRouter(<Guessing stage={stage} test={true} />);
    const correctAnswer = await waitForElement(() => getByText("Chair", { selector: 'button' }));

    // console.error = jest.fn()

    await act(async () => {
        fireEvent.click(correctAnswer);
    });

    const option1 = await waitForElement(() => getByText("Bear", { selector: 'button' }));
    const option2 = await waitForElement(() => getByText("Bed", { selector: 'button' }));
    const option3 = await waitForElement(() => getByText("Chair", { selector: 'button' }));
    const option4 = await waitForElement(() => getByText("Porridge", { selector: 'button' }));
    expect(option1).toBeDisabled();
    expect(option2).toBeDisabled();
    expect(option3).toBeDisabled();
    expect(option4).toBeDisabled();

    const nextButton = await waitForElement(() => getByTestId('next-button'));
    expect(nextButton).toBeTruthy();
})

it('displays correct ui after inccorect answer is selected', async () => {
    const { getByTestId, getByText } = renderWithReduxRouter(<Guessing stage={stage} />);
    const incorrectAnswer = await waitForElement(() => getByText("Bear", { selector: 'button' }));

    await act(async () => {
        fireEvent.click(incorrectAnswer);
    });

    const option1 = await waitForElement(() => getByText("Bear", { selector: 'button' }));
    const option2 = await waitForElement(() => getByText("Bed", { selector: 'button' }));
    const option3 = await waitForElement(() => getByText("Chair", { selector: 'button' }));
    const option4 = await waitForElement(() => getByText("Porridge", { selector: 'button' }));
    expect(option1).toBeDisabled();
    expect(option2).not.toBeDisabled();
    expect(option3).not.toBeDisabled();
    expect(option4).not.toBeDisabled();

    expect(() => getByTestId("next-button")).toThrow('Unable to find an element by: [data-testid="next-button"]');
})