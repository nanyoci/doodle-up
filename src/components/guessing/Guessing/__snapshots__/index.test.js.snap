// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays correct ui after correct answer is selected 1`] = `
<div>
  <div
    class="main-content"
  >
    <audio />
    <div
      class="wrapper"
    >
      <a
        href="/"
      >
        <img
          alt="DoodleUp"
          class="logo-corner"
          src="doodle-up.svg"
        />
      </a>
      <div
        class="guessing-container"
      >
        <h1
          id="guessTheName"
        >
          Guess the Name!
        </h1>
        <div
          id="guessingImgDiv"
        >
          <img
            alt="For guessing"
            id="guessingImg"
            src="drawing-sample.png"
          />
        </div>
        <div
          class="options"
        >
          <div
            class="option-unit"
          >
            <button
              alt="audio"
              class="ui violet large circular icon button"
            >
              <i
                aria-hidden="true"
                class="volume up icon"
              />
            </button>
            <button
              class="option btn btn-outline-dark"
            >
              Bear
            </button>
          </div>
          <div
            class="option-unit"
          >
            <button
              alt="audio"
              class="ui violet large circular icon button"
            >
              <i
                aria-hidden="true"
                class="volume up icon"
              />
            </button>
            <button
              class="option btn btn-outline-dark"
            >
              Bed
            </button>
          </div>
          <div
            class="option-unit"
          >
            <button
              alt="audio"
              class="ui violet large circular icon button"
            >
              <i
                aria-hidden="true"
                class="volume up icon"
              />
            </button>
            <button
              class="option btn btn-danger"
              disabled=""
            >
              Chair
            </button>
          </div>
          <div
            class="option-unit"
          >
            <button
              alt="audio"
              class="ui violet large circular icon button"
            >
              <i
                aria-hidden="true"
                class="volume up icon"
              />
            </button>
            <button
              class="option btn btn-outline-dark"
            >
              Porridge
            </button>
          </div>
        </div>
      </div>
      <div
        class="reward-container"
      >
        <div
          class="rewards"
        >
          <div />
          <div
            style="position: relative;"
          />
          <div
            style="transform: none;"
          />
        </div>
        <div
          class="rewards"
        >
          <div />
          <div
            style="position: relative;"
          />
          <div
            style="transform: none;"
          />
        </div>
      </div>
    </div>
    <div
      class="control-buttons"
    >
      <button
        class="ui yellow massive circular icon button"
        id="signOutBtn"
      >
        <i
          aria-hidden="true"
          class="sign-out icon"
        />
      </button>
      <button
        class="ui yellow massive circular icon button"
        id="musicControlBtn"
      >
        <i
          aria-hidden="true"
          class="mute icon"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="main-content"
  >
    <audio />
    <div
      class="wrapper"
    >
      <a
        href="/"
      >
        <img
          alt="DoodleUp"
          class="logo-corner"
          src="doodle-up.svg"
        />
      </a>
      <div
        class="guessing-container"
      >
        <h1
          id="guessTheName"
        >
          Guess the Name!
        </h1>
        <div
          id="guessingImgDiv"
        >
          <img
            alt="For guessing"
            id="guessingImg"
            src="drawing-sample.png"
          />
        </div>
        <div
          class="options"
        >
          <div
            class="option-unit"
          >
            <button
              alt="audio"
              class="ui violet large circular icon button"
            >
              <i
                aria-hidden="true"
                class="volume up icon"
              />
            </button>
            <button
              class="option btn btn-outline-dark"
            >
              Bear
            </button>
          </div>
          <div
            class="option-unit"
          >
            <button
              alt="audio"
              class="ui violet large circular icon button"
            >
              <i
                aria-hidden="true"
                class="volume up icon"
              />
            </button>
            <button
              class="option btn btn-outline-dark"
            >
              Bed
            </button>
          </div>
          <div
            class="option-unit"
          >
            <button
              alt="audio"
              class="ui violet large circular icon button"
            >
              <i
                aria-hidden="true"
                class="volume up icon"
              />
            </button>
            <button
              class="option btn btn-outline-dark"
            >
              Chair
            </button>
          </div>
          <div
            class="option-unit"
          >
            <button
              alt="audio"
              class="ui violet large circular icon button"
            >
              <i
                aria-hidden="true"
                class="volume up icon"
              />
            </button>
            <button
              class="option btn btn-outline-dark"
            >
              Porridge
            </button>
          </div>
        </div>
      </div>
      <div
        class="reward-container"
      >
        <div
          class="rewards"
        >
          <div />
          <div
            style="position: relative;"
          />
          <div
            style="transform: none;"
          />
        </div>
        <div
          class="rewards"
        >
          <div />
          <div
            style="position: relative;"
          />
          <div
            style="transform: none;"
          />
        </div>
      </div>
    </div>
    <div
      class="control-buttons"
    >
      <button
        class="ui yellow massive circular icon button"
        id="signOutBtn"
      >
        <i
          aria-hidden="true"
          class="sign-out icon"
        />
      </button>
      <button
        class="ui yellow massive circular icon button"
        id="musicControlBtn"
      >
        <i
          aria-hidden="true"
          class="mute icon"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
